Change dialog to position:fixed, and remove dialog alignment modes.

bfgeek opened the pull request:
- [ ] At least two implementers are interested (and none opposed):
   * https://github.com/w3c/csswg-drafts/issues/4645
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * Tests will be updated in https://chromium-review.googlesource.com/c/chromium/src/+/2258897
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Firefox: …
   * Safari: …

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5936/infrastructure.html" title="Last updated on Oct 29, 2020, 10:05 PM UTC (a75a0d8)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/5936/cc2ba62...a75a0d8/infrastructure.html" title="Last updated on Oct 29, 2020, 10:05 PM UTC (a75a0d8)">diff</a> )
<a href="https://whatpr.org/html/5936/interactive-elements.html" title="Last updated on Oct 29, 2020, 10:05 PM UTC (a75a0d8)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/5936/cc2ba62...a75a0d8/interactive-elements.html" title="Last updated on Oct 29, 2020, 10:05 PM UTC (a75a0d8)">diff</a> )
<a href="https://whatpr.org/html/5936/rendering.html" title="Last updated on Oct 29, 2020, 10:05 PM UTC (a75a0d8)">/rendering.html</a>  ( <a href="https://whatpr.org/html/5936/cc2ba62...a75a0d8/rendering.html" title="Last updated on Oct 29, 2020, 10:05 PM UTC (a75a0d8)">diff</a> )

domenic commented on source:
```diff
@@ -57253,8 +57251,6 @@ interface <dfn>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
 
    <li><p>Add an <code data-x="attr-dialog-open">open</code> attribute to <var>subject</var>, whose value is the empty string.</p></li>
 
-   <li><p>Set the <code>dialog</code> to the <span>centered alignment</span> mode.</p></li>
```
So is the intention really that show() and showModal() now have the same visual behavior? That's a bit surprising, but if it's what everyone wants...

dpogue replied:
I agree this behaviour seems surprising.
It makes sense for `showModal()` to use `position: fixed`, but I wouldn't expect that behaviour when calling `show()`.

At risk of this ended up back in the CSSWG, it feels like what would make sense is something like this:
```css
dialog {
  position: absolute;
}

dialog:modal {
  position: fixed;
}
```

bfgeek replied:
@sefeng211 - Thoughts on above? I'd be happy with the above proposal where the applied CSS would be:

```css
dialog:modal {
  position: fixed;
  overflow: auto;
  inset-block-start: 0; inset-block-end: 0;
  /* 6px + 2em = border + padding */
  max-width: calc(100% - 6px - 2em);
  max-height: calc(100% - 6px - 2em);
}
```

sefeng211 replied:
I agree. Make dialog to be `absolute` and modal dialog to be `fixed` looks much safer!

domenic commented on source:
```diff
@@ -113090,10 +113041,14 @@ listing, plaintext, pre, xmp {
   height: fit-content;
+  /* 6px + 2em = border + padding */
+  max-width: calc(100% - 6px - 2em);
+  max-height: calc(100% - 6px - 2em);
```
Today I learned the default border width is 3px; fun!

bfgeek replied:
Yup! In an ideal world this would have `box-sizing: border-box` applied, but this would likely change existing users of `<dialog>` a little too much.

emilio replied:
So:

 * I agree that the fixed-pos thing should only be done for modal dialogs.

 * In Gecko, we're using `max-width: 100vw; max-height: 100vh;` that seems better than this, as it preserves the right max size when authors change borders / padding, or `box-sizing`, which I expect to be common

 * Is `block-size: fit-content` well defined? In Gecko, it just behaves like auto, so an slightly more expressive way to do this would be `inline-size: fit-content; block-size: auto;` (or omitting block-size altogether). I guess that can be a separate discussion, but I think it is relevant because if it is specified to behave like auto, or it is not well defined on the block axis, then the `top: 0; bottom: 0` will stretch the dialog, I _think_ (late, so double-check me here Ian). In Gecko we do `top: 50%; transform: translateY(-50%);`, which conveniently is how this is done in the wild in all the examples we found, so it doesn't break positioning by sites, see https://bugzilla.mozilla.org/show_bug.cgi?id=1650820.

For reference, our stylesheet is [this](https://searchfox.org/mozilla-central/rev/3d53187b90605ccef321c9cadbba762ad06a6381/layout/style/res/html.css#803-848) as of today. Modulo that bug I linked above (which is what caused us to use `transform` rather than `translate`, and which is fixed except for a UA-sniffing thing they're doing), we haven't found any other compat issues since we enabled it by default on Nightly months ago.

bfgeek replied:
> In Gecko, we're using max-width: 100vw; max-height: 100vh; that seems better than this, as it preserves the right max size when authors change borders / padding, or box-sizing, which I expect to be common

The `max-height` is incorrect on FF today by default however, e.g. the border/padding/scrollbar is clipped when the size is exceeded.

> Is block-size: fit-content well defined?

`block-size: fit-content` - This might be something that is generally underspecified.

This is used to trigger centering based on margins, as you've specified a height. This is generally useful for web-developers maybe we should file a bug on css-sizing / css-position to actually specify that this is supported? (To be clear I'm against us changing behaviour here to *un-support* this :) )

I'm also opposed to using `transform` as you've done this would break/prevent any sites using `transform` to animate a dialog in/out etc.

bfgeek replied:
> block-size: fit-content - This might be something that is generally underspecified.

Quickly looking at css-position you could make the argument that Blink's behaviour is expected.

emilio replied:
> I'm also opposed to using transform as you've done this would break/prevent any sites using transform to animate a dialog in/out etc.

How so? The transform is overriden by the authors. These are not `!important` styles.

emilio replied:
> The max-height is incorrect on FF today by default however, e.g. the border/padding/scrollbar is clipped when the size is exceeded.

Ah, that's right... hm, I guess we either need to add `box-sizing: border-box`, or subtract border/padding. I think doing border-box sizing is cleaner, but if you'd rather leave the calc() that's probably fine too.

bfgeek replied:
> Ah, that's right... hm, I guess we either need to add box-sizing: border-box, or subtract border/padding. I think doing border-box sizing is cleaner, but if you'd rather leave the calc() that's probably fine too.

Right - the change to `border-box` will likely break too many things.

> How so? The transform is overriden by the authors. These are not !important styles.

This will break existing animations, e.g. if I've specified an animation from `translate(-1000px)` -> `translate(0)` this will look broken in FF as they don't accumulate.

fantasai replied:
I think you want[ `max-width/height: stretch` ](https://www.w3.org/TR/css-sizing-4/#valdef-width-stretch)?

domenic reviewed (approved):
Looks pretty solid. @sefeng211 could you take a look too?

Also please make sure there are tests for the computedStyle of an unstyled dialog matching the spec's UA stylesheet. https://github.com/web-platform-tests/wpt/issues/5625 discusses the general problem and solution space (/cc @zcorpan) but for now some one-off tests would be good.

sefeng211 reviewed (approved):
Looks good!

emilio commented on source:
```diff
@@ -113090,10 +113041,14 @@ listing, plaintext, pre, xmp {
-  inset-inline-start: 0; inset-inline-end: 0;
+  position: fixed;
+  overflow: auto;
+  top: 0; right: 0; bottom: 0; left: 0;
```
this can be written as `inset: 0;`

emilio commented:
So regarding changing the styling of non-modal dialogs... I think I don't have a super-strong opinion. I think those right now don't have any specialness (other than the styling defined on the spec), so it may be less risky to just change the modal ones to be fixed.

bfgeek commented:
> So regarding changing the styling of non-modal dialogs... I think I don't have a super-strong opinion. I think those right now don't have any specialness (other than the styling defined on the spec), so it may be less risky to just change the modal ones to be fixed.

We'll need a `:modal` pseudo class in that case, (unless you are suggesting that we write to the inline-styles or something?)

emilio commented:
> We'll need a :modal pseudo class in that case, (unless you are suggesting that we write to the inline-styles or something?)

Yeah, that's effectively what we have internally in Gecko. Not sure if such a pseudo-class needs to be exposed to the web though.

bfgeek commented:
> Yeah, that's effectively what we have internally in Gecko. Not sure if such a pseudo-class needs to be exposed to the web though.

I'd slightly prefer to expose this to web-developers as 1) allows greater control, and 2) doesn't seem to be a reason not to?

bfgeek commented:
@emilio - Are you ok w/ the `height: fit-content` (and I'll file an issue on the CSSWG spec clarifying our behaviour?) - and OK w/ the additional of a `:modal` pseudo?

emilio commented:
Yes, I'd be ok with those. And per our other discussion it should be fine to make the max/min-height `calc(100% - borderpadding)` rather than viewport units / box-size.

bfgeek commented:
@sefeng211 @emilio @domenic - I've added the `:modal` pseudo-class which relies on a flag within each `dialog` to conditionally set the `position: fixed` etc styles, PTAL :) .

domenic reviewed (dismissed):
LGTM, editorially and matches my expectations. I'm happy to merge after @emilio confirms. It seems like @sefeng211 already has chimed in to agree with this version.

sefeng211 reviewed (approved):
LGTM!

emilio reviewed (approved):
Looks good, thanks!

annevk commented:
@whatwg/css do new pseudo-classes need to be coordinated with the CSS WG?

(Apologies if I missed something, but I don't see it in https://drafts.csswg.org/selectors/ and it's not discussed in https://github.com/w3c/csswg-drafts/issues/4645 either as far as I can tell.)

bfgeek commented:
> @whatwg/css do new pseudo-classes need to be coordinated with the CSS WG?

Not typically - many examples of pseudo-classes being defined outside of css specifications, e.g.
Many of the pseudo-classes in the html spec, e.g. https://html.spec.whatwg.org/#selector-in-range
https://immersive-web.github.io/dom-overlays/#xr-overlay
etc.

annevk commented:
I'm not familiar with XR, but https://drafts.csswg.org/selectors/#range-pseudos covers the one you mentioned in HTML. And I'm pretty sure the CSS WG wanted to be involved in any additions going forward as HTML minting a couple came as a surprise. (It also seems there needs to be some involvement as there is no registry other than the Selectors document.)

bfgeek commented on source:
```diff
@@ -113100,6 +113068,14 @@ dialog {
+  inset-block-start: 0; inset-block-end: 0;
+  /* 6px + 2em = border + padding */
+  max-width: calc(100% - 6px - 2em);
+  max-height: calc(100% - 6px - 2em);
```
I'm fine w/ this if @emilio is. Change pre-emptively to 'stretch'.

annevk replied:
@domenic not sure if you noticed this but @emilio pointed out to me that currently no browser implements stretch. Firefox might eventually, but if someone were to write a test about the computed value for a dialog element that is modal, they would not see stretch...

domenic replied:
Hmm. Will browsers block implementing dialog until they implement stretch then? Or should we change this to something implementable?

annevk replied:
I don't think Firefox would block on `stretch`.

domenic replied:
What will Firefox implement instead? (Same question for Blink.)

annevk replied:
```
  max-width: calc(100% - 6px - 2em);
  max-height: calc(100% - 6px - 2em);
```
as it was before. Once we get around to supporting `stretch` we'd be up for changing this, but we don't necessarily want to block on it. If Chrome does `stretch` though it seems fine to keep the PR as-is.

bfgeek replied:
We don't have `stretch` but would use `-webkit-fill-available` which we've been slowly been moving to the defined `stretch` behaviour. For this particular case it is a drop in replacement. However happy also if we keep as the `calc(100% - 6px - 2em)`.

domenic replied:
Hmm. So I guess we have two concerns here:

1. Interop. If Blink uses `-webkit-fill-available` and Gecko uses `calc()`, it sounds like there will be interop problems if the author changes the border width. Is this OK?

2. Compat. If both engines use `calc()`, then we might be prevented from ever upgrading to `stretch` since it would impact the layout of `<dialog>`s whose border width was changed.

These trade off against each other, e.g. by sacrificing interop, we can preserve a bit of flexibility for future changes.

If we prioritize interop, then I'd say the spec (and Blink) should probably use `calc()`, and in the future, when both engines implement `stretch`, we could consider upgrading the spec, subject to the usual slow change process for any compat-impacting upgrade of this sort.

If we prioritize compat, then the spec should probably use `stretch`, since it indicates the desired behavior and roughly matches Blink's `-webkit-fill-available` IIUC.

(There's a third concern, roughly "conformance", which is that it'd sure be nice if everyone agreed with the spec on things like `getComputedStyle(dialog).maxWidth`. But that's pretty low priority.)

bfgeek replied:
@emilio Thoughts?

emilio replied:
For the inline axis we could probably use `-moz-available`. We don't have anything hooked up for the block axis to do this, I think, but it may not be too hard to either hook that or implement `stretch`.

Or we can probably just use `calc()`, honestly I think it's ~fine.

This effectively only changes the total max size of the dialog which is intended to prevent unreachable dialogs. It is mostly a safety mechanism. I don't think we'd have much trouble changing it to `stretch` in the future (because an overflowing modal dialog is basically always a bug, and `stretch` will be either equivalent or superior to both solutions).

I think the most concerning case against `calc()` is full-width dialog in the inline-axis which the author has styled to use less padding/border (or `box-sizing: border-box`). In that case `calc()` will incorrectly shrink the dialog too much in the inline axis and it could look a bit ugly (I'm not _too_ concerned for shrinking too much in the block axis I think...).

But generally it seems both Chromium and Gecko have a solution for the inline axis that does the same as `stretch`, and I don't think using either solution prevents us from moving to `stretch` in the future.

bfgeek replied:
I agree it likely won't prevent us from moving to `stretch` in the future, I've changed back to the `calc()` version.

bfgeek commented on source:
```diff
@@ -113100,6 +113068,14 @@ dialog {
+dialog:modal {
+  position: fixed;
+  overflow: auto;
+  inset-block-start: 0; inset-block-end: 0;
```
They are necessary as the default styles only include the inline direction variants.

bfgeek commented on source:
```diff
@@ -113100,6 +113068,14 @@ dialog {
 }
+dialog:modal {
+  position: fixed;
+  overflow: auto;
```
I don't think so?

anthony-redFox replied:
@fserb Can you explain why for `dialog:modal` was added `overflow: auto`?

That mean if dialog has html dropdowns or html tooltips which shows out of the dialog are broken because dialog shows scrolls by default. Can we eliminate scrolls in dialogs at all by default?

cc @sefeng211 @domenic

emilio replied:
I think scroll by default makes sense, otherwise dialogs that are larger than the viewport are completely unreadable. If you're 100% sure that your content won't overflow the viewport you can set `overflow: visible`.

anthony-redFox replied:
It is rare case when dialog has the same size as viewport, any case the change was break us. And question that Do we have strong reason to have `overflow: auto` (Ideally - to remove it.).

emilio replied:
It could happen fairly easily if you have a small viewport and large content in the dialog. Defaulting to not leaving content unreachable makes sense imo.

bfgeek commented:
@emilio @domenic @annevk I've change the pseudo to `:-internal-modal` add added prose that should should not be exposed outside the UA stylesheet. We can describe the `dialog:-internal-modal` rule in prose instead, but if describing these things in prose has taught us anything it'll likely be incorrect :).

domenic commented:
I don't think we should definite an internal pseudo-class for this, if it's not observable by web content. Instead I'd suggest using prose. E.g. like is done in other sentences in the Rendering section like

> When an input element's dir attribute is in the auto state and its type attribute is in the Text state, then the user agent is expected to act as if it had a user-agent-level style sheet rule setting the 'unicode-bidi' property to 'plaintext'.

or

> When the document's character encoding is ISO-8859-8, the following rules are additionally expected to apply, following those above:

domenic commented:
Oh, comment collision. Yeah, I do prefer prose a good amount, and I think it aligns better with precedent.

bfgeek commented:
> I don't think we should definite an internal pseudo-class for this, if it's not observable by web content. Instead I'd suggest using prose. E.g. like is done in other sentences in the Rendering section like

Even if this is how all the UAs do this?

domenic commented:
Yeah. I think although it is convenient for UAs to mix together internal implementation details with the author-facing pseudo-class concept, it's not a great idea at the spec level, and muddies the meaning of pseudo-classes.

If the CSSWG would like to formally define a notion of internal pseudo-classes, and if we did a wholesale update of all similar cases in HTML (e.g. an `:-internal-ISO-8859-8` pseudo-class), then it would be a reasonable route. But until then, I think they should be in prose.

bfgeek commented:
> If the CSSWG would like to formally define a notion of internal pseudo-classes

<fantasai> RESOLVED: Add a hidden pseudo-class for this purpose, in order to solve styling issue while leavine open the possibility of HTML improving its API

?

bfgeek commented:
(from https://github.com/w3c/csswg-drafts/issues/4645 )

domenic commented:
Oh OK, so you'll also work on defining the internal pseudo-class concept in CSS? It feels like that would involve changes to the parser and cascade specs, right?

domenic commented on source:
```diff
@@ -57371,48 +57376,20 @@ interface <dfn>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
-  element is larger than the <span>viewport</span> and the <span>viewport</span> would otherwise not
-  have a scroll mechanism (e.g. because the <span>viewport</span>'s <span>'overflow'</span> property
-  is set to <span>'hidden'</span>).</p>
+  <p>When this flag is true the user agent is expected to act as if it had a user-agent-level style
```
Hmm. I think we should probably centralize this in the rendering section with other rendering changes?

bfgeek replied:
Done.

domenic reviewed (approved):
LGTM with potential nit

domenic reviewed (approved):
Continues to LGTM. Will give this another day or two in case folks want to review further, but I'm looking forward to merging this.

annevk commented:
https://github.com/whatwg/html/pull/5936#discussion_r510542199 is probably worth looking into once more before merging.

annevk commented:
Could someone update OP with the implementation bugs?

bfgeek commented:
> Could someone update OP with the implementation bugs?

Done - @emilio I used https://github.com/mozilla/wg-decisions/issues/412 lmk if you'd like something else.

emilio commented:
https://bugzilla.mozilla.org/show_bug.cgi?id=1675857, thanks @bfgeek
